sonar-text-plugin/src/main/java/org/sonar/plugins/text/checks/BIDICharacterCheck.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * SonarQube Text Plugin
+ * Copyright (C) 2021-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.text.checks;
+
+import java.io.InputStream;
+import java.util.List;
+import java.util.Scanner;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.check.Rule;
+import org.sonar.plugins.text.api.CheckContext;
+
+@Rule(key = "placeholder")
+public class BIDICharacterCheck extends AbstractInputStreamCheck {
+
+  public static final String MESSAGE = "Make sure that using bidirectional characters is safe here.";
+  private static final List<Character> BIDI_CHARS = List.of(
+    '\u061C',
+    '\u200E',
+    '\u200F',
+    '\u202A',
+    '\u202B',
+    '\u202C',
+    '\u202D',
+    '\u202E',
+    '\u2066',
+    '\u2067',
+    '\u2068',
+    '\u2069'
+  );
+
+  @Override
+  void analyzeStream(CheckContext ctx, InputFile inputFile, InputStream stream) {
+    Scanner scanner = new Scanner(stream, inputFile.charset().name());
+    int lineNumber = 0;
+    while (scanner.hasNextLine()) {
+      lineNumber++;
+      checkLine(ctx, scanner.nextLine(), lineNumber);
+    }
+  }
+
+  private static void checkLine(CheckContext ctx, String lineContent, int lineNumber) {
+    for (Character bidiChar : BIDI_CHARS) {
+      if (lineContent.indexOf(bidiChar) >= 0) {
+        ctx.reportLineIssue(lineNumber, MESSAGE);
+      }
+    }
```

Comment:
Wouldn't it be faster to have a regex with a character class looking for a BIDI char in the line?
```suggestion
    if (BIDI_CHARS_PATTER.matcher(lineContent).matches()) {
      ctx.reportLineIssue(lineNumber, MESSAGE);
    }
```
with
```
  private static final Pattern BIDI_CHARS_PATTER;
  static {
    List<String> chars = BIDI_CHARS.stream().map(String::valueOf).collect(Collectors.toList());
    BIDI_CHARS_PATTER = Pattern.compile(String.format("[%s]", String.join("|", chars)));
  }
```
WDYT?

Reply:
If not, I would stop the loop on a match
```suggestion
    for (Character bidiChar : BIDI_CHARS) {
      if (lineContent.indexOf(bidiChar) >= 0) {
        ctx.reportLineIssue(lineNumber, MESSAGE);
        return;
      }
    }
```

Reply:
I did try it out with regex as you proposed now, and the characters were not detected in the test.
Not sure why, but before investigating I did measure if it makes a difference on big files (I took tcpdf with its around 26k lines), and both approaches were more or less the same (regex a bit slower), but we are talking in 10th of milliseconds here (e.g., 98ms vs 108ms), so I don't think it's worth investigating now.